File: work/csv/JsonToCsv.java
package com.codenjoy.dojo.utils.csv;

import com.codenjoy.dojo.utils.JsonUtils;
import lombok.SneakyThrows;

import java.util.List;
import java.util.Map;

import static com.codenjoy.dojo.utils.JsonUtils.SORT_KEYS;
import static com.codenjoy.dojo.utils.csv.Matrix.SEPARATOR;

public class JsonToCsv {

    private static final boolean DEBUG = false;

    public static String jsonToCsv(String json) {
        return jsonToCsv(json,
                ",",
                "\"",
                "\"",
                "\n\",",
                null,
                List.of("\"" + SEPARATOR + "\"\""),
                false,
                true);
    }

    public static String jsonToMarkdown(String json) {
        return jsonToCsv(json,
                "|",
                null,
                null,
                null,
                "-",
                null,
                true,
                true);
    }

    @SneakyThrows
    public static String jsonToCsv(
            String json,
            String columnDelimiter,
            String cellLeftDelimiter,
            String cellRightDelimiter,
            String lineCharsNeedToBeEscapedWithCellDelimiter,
            String headerDelimiter,
            List<String> lineReplacements,
            boolean paddingToMaxCellLength,
            boolean removeHeadersDuplicates
    ) {
        List<Object> data = JsonUtils.jsonMapper(!SORT_KEYS).readValue(json, List.class);

        Matrix matrix = new Matrix();

        processElement(matrix, data, "", false, 0);

        if (removeHeadersDuplicates) {
            matrix.removeHeadersDuplicates();
        }
        return matrix.toString(
                columnDelimiter,
                cellLeftDelimiter,
                cellRightDelimiter,
                lineCharsNeedToBeEscapedWithCellDelimiter,
                headerDelimiter,
                lineReplacements,
                paddingToMaxCellLength);
    }

    private static void processElement(Matrix matrix, Object obj, String prefix, boolean sameLine, int depth) {
        int x = 0;
        if (!prefix.isEmpty()) {
            x = matrix.getOrCreateHeader(prefix);
        }
        if (obj instanceof Map) {
            Map<String, Object> map = (Map<String, Object>) obj;
            if (depth > 1) {
                matrix.childHeader();
            }
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                String key = prefix.isEmpty() ? entry.getKey() : prefix + "." + entry.getKey();
                Object value = entry.getValue();
                processElement(matrix, value, key, sameLine, depth + 1);
                sameLine = true;
            }
            if (depth > 1) {
                matrix.parentHeader();
            }
        } else if (obj instanceof List) {
            List<Object> list = (List<Object>) obj;
            for (int y = 0; y < list.size(); y++) {
                if (depth == 0) {
                    matrix.newLine();
                } else {
                    sameLine = y == 0;
                }
                Object item = list.get(y);
                processElement(matrix, item, prefix, sameLine, depth + 1);
            }
        } else {
            matrix.inject(x, sameLine, String.valueOf(obj));
            if (DEBUG) {
                System.out.println("Injecting: " + prefix + " -> " + obj);
                System.out.println("===============================================================================");
                System.out.println(matrix.toString("|", null, null, null, "-", null, true));
                System.out.println("===============================================================================");
            }
        }
    }
}

File: mcp_server/tools/lng_json_to_csv/stuff/json_to_csv.py
"""Python port of JsonToCsv.java"""

import json
from typing import List, Optional, Any, Dict, Union
from matrix import Matrix


class JsonToCsv:
    """Python port of JsonToCsv.java"""
    
    DEBUG = False
    
    @staticmethod
    def json_to_csv(json_str: str) -> str:
        """Convert JSON to CSV with default parameters"""
        return JsonToCsv._json_to_csv(
            json_str,
            ",",
            '"',
            '"',
            '\n",',
            None,
            ['"' + Matrix.SEPARATOR + '""'],
            False,
            True
        )
    
    @staticmethod
    def json_to_markdown(json_str: str) -> str:
        """Convert JSON to Markdown table format"""
        return JsonToCsv._json_to_csv(
            json_str,
            "|",
            None,
            None,
            None,
            "-",
            None,
            True,
            True
        )
    
    @staticmethod
    def _json_to_csv(
        json_str: str,
        column_delimiter: str,
        cell_left_delimiter: Optional[str],
        cell_right_delimiter: Optional[str],
        line_chars_need_to_be_escaped_with_cell_delimiter: Optional[str],
        header_delimiter: Optional[str],
        line_replacements: Optional[List[str]],
        padding_to_max_cell_length: bool,
        remove_headers_duplicates: bool
    ) -> str:
        """Internal method to convert JSON to CSV/Markdown"""
        
        # Parse JSON data
        data = json.loads(json_str)
        
        # Ensure data is a list
        if not isinstance(data, list):
            data = [data]
        
        matrix = Matrix()
        
        # Process each element in the data
        JsonToCsv._process_element(matrix, data, "", False, 0)
        
        # Remove header duplicates if requested
        if remove_headers_duplicates:
            matrix.remove_headers_duplicates()
        
        # Convert matrix to string
        return matrix.to_string(
            column_delimiter,
            cell_left_delimiter,
            cell_right_delimiter,
            line_chars_need_to_be_escaped_with_cell_delimiter,
            header_delimiter,
            line_replacements,
            padding_to_max_cell_length
        )
    
    @staticmethod
    @staticmethod
    def _process_element(matrix: Matrix, obj: Any, prefix: str, same_line: bool, depth: int):
        """Process individual elements recursively"""
        x = 0
        if prefix:
            x = matrix.get_or_create_header(prefix)
        
        if isinstance(obj, dict):
            # Handle dictionary/object
            if depth > 1:
                matrix.child_header()
            
            # Use original dict order (insertion order in Python 3.7+)
            for key, value in obj.items():
                new_key = key if not prefix else f"{prefix}.{key}"
                JsonToCsv._process_element(matrix, value, new_key, same_line, depth + 1)
                same_line = True
            
            if depth > 1:
                matrix.parent_header()
        
        elif isinstance(obj, list):
            # Handle array/list
            for i, item in enumerate(obj):
                if depth == 0:
                    matrix.new_line()
                else:
                    same_line = i == 0
                
                JsonToCsv._process_element(matrix, item, prefix, same_line, depth + 1)
        
        else:
            # Handle primitive values
            # Convert boolean values to lowercase like Java
            if isinstance(obj, bool):
                value = "true" if obj else "false"
            else:
                value = str(obj)
            matrix.inject(x, same_line, value)
            
            if JsonToCsv.DEBUG:
                print(f"Injecting: {prefix} -> {obj}")
                print("=" * 80)
                print(matrix.to_string("|", None, None, None, "-", None, True))
                print("=" * 80)
                print(matrix.to_string("|", None, None, None, "-", None, True))
                print("=" * 80)
